static/src/javascripts/projects/common/modules/navigation/newSticky.js

```diff
@@ -28,68 +28,112 @@ define([
-
-                            fastdom.write(function () {
-                                // Reset
-                                $header.css('transition', '');
-                                $adBanner.css('top', '');
-
-                                // Set
-                                if (scrollY > headerHeight) {
-                                    $adBanner.css({
-                                        'position': 'absolute',
-                                        'top': headerHeight + 'px'
-                                    });
-                                } else {
-                                    $adBanner.css({ 'position': 'fixed' });
-                                }
-
-                                $header.css('margin-top', adHeight + 'px');
-
-                                var scrollIsAtTop = scrollY === 0;
-                                // Avoid an initial transition when we apply the margin top for the first time
-                                if (scrollIsAtTop && !firstRender) {
-                                    // If the user is at the top of the page, we want to transition
-                                    // the change
-                                    $header.css('transition', 'margin-top 0.75s cubic-bezier(0, 0, 0, 0.985)');
-                                } else {
-                                    // If the user is not at the top, we want to offset their scroll position
-                                    var diff = adHeight - previousAdHeight;
-                                    if (diff > 0) {
-                                        window.scrollTo(scrollX, scrollY + diff);
-                                    }
-                                }
+                var getLatestAdHeight = function () {
+                    var $iframe = $('iframe', $adBanner);
+                    var slotWidth = $iframe.attr('width');
+                    var slotHeight = $iframe.attr('height');
+                    // iframe may not have been injected at this point
+                    var isFluidAd = $iframe.length > 0 && [slotWidth, slotHeight].join(',') === '88,70';
+                    // fluid ads are currently always 250px high. We can't just read the client height as fluid ads are
+                    // injected asynchronously, so we can't be sure when they will be in the dom
+                    var fluidAdInnerHeight = 250;
+                    var fluidAdPadding = 18;
+                    var fluidAdHeight = fluidAdInnerHeight + fluidAdPadding;
 
-                                previousAdHeight = adHeight;
+                    return isFluidAd
+                        ? Promise.resolve(fluidAdHeight)
+                        : getClientAdHeight();
+                };
+
+                var oldAdHeightPromise = getLatestAdHeight();
+                var newAdHeightPromise = topAdRenderedPromise.then(getLatestAdHeight);
+
+                var getCachedAdHeight = function () {
```

Comment:
This is an interesting thing to discuss. Imagine you have a promise for a user. If we called this `user`, you would either have to shadow the identifier in the `then` callback, or call it something else:

``` js
user.then(user => {
  // do something with user
});
```

I want to avoid doing this because shadowing hurts readability, so the `<identifier>Promise` is just one way of doing that. Strings aren't wrapped values, they are just values themselves, so they don't have this problem.

We have the same problem with futures and options in Scala, and it is common convention to prefix identifiers with their wrapped type, i.e. `maybeUser` for options or `futureUser` for futures.

--

In the example above, the value of `getLatestAdHeight` is a function which returns a promise. I think we only need the above convention when dealing with promise values themselves.